src/main/java/com/palantir/refreshable/DefaultRefreshable.java

```diff
@@ -84,13 +85,18 @@ private DefaultRefreshable(T current, Optional<?> strongParentReference, RootSub
 
     /** Updates the current value and sends the specified value to all subscribers. */
     @Override
-    public synchronized void update(T value) {
-        if (!Objects.equals(current, value)) {
-            current = value;
-
-            // iterating over a copy allows SelfRemovingSubscribers to remove themselves without
-            // ConcurrentModificationExceptions
-            ImmutableList.copyOf(orderedSubscribers).forEach(subscriber -> subscriber.accept(value));
+    public void update(T value) {
+        writeLock.lock();
+        try {
+            if (!Objects.equals(current, value)) {
+                current = value;
+
+                // iterating over a copy allows subscriptions to be disposed within an update without causing
+                // ConcurrentModificationExceptions.
+                ImmutableList.copyOf(orderedSubscribers).forEach(subscriber -> subscriber.accept(value));
```

Comment:
I believe you are behind an exclusive lock, so don't need the copy?

Reply:
There’s a test which shows why this is important. On update a subscriber may dispose of another subscription, so I’d prefer to avoid the entire class of problem.

the exclusive lock doesn’t prevent reentrant accesses on the same thread from modifying the collection within iteration.

Reply:
The case is not very likely, however I’d rather make updates a little slower than break services that have been happily running with spaghetti-shaped subscriptions.